src/platform/packages/shared/kbn-unified-metrics-grid/src/context/metrics_experience_fields_provider/helpers/fields_parser.ts

```diff
@@ -0,0 +1,200 @@
+  '_tier',
+  '_metric_names_hash',
+]);
+
+const shouldSkipField = (fieldName: string) => FILTER_OUT_FIELDS.has(fieldName);
+
+/**
+ * Extracts metric fields and dimensions from a data view and esql result columns.
+ * @param index - The index pattern to use for the metric fields.
+ * @param dataViewFieldMap - The data view field map to use for the metric fields.
+ * @param columns - The columns to use for the metric fields.
+ * @returns The metric fields and dimensions.
+ */
+export const extractFields = ({
+  index,
+  dataViewFieldMap,
+  columns = [],
+}: {
+  index: string;
+  dataViewFieldMap: DataViewFieldMap;
+  columns?: DatatableColumn[];
+}): {
+  metricFields: MetricField[];
+  dimensions: Dimension[];
+} => {
+  const metricFields: MetricField[] = [];
+  const dimensions: Dimension[] = [];
+
+  const columnByName = new Map<string, DatatableColumn>(
+    columns?.map((column) => [column.name, column]) ?? []
+  );
+
+  for (const [columnName, column] of columnByName) {
+    if (shouldSkipField(columnName) || !Object.hasOwn(dataViewFieldMap, columnName)) {
+      continue;
+    }
+
+    const dataViewField = dataViewFieldMap[columnName];
+    const fieldType = (column?.meta?.esType || dataViewField.esTypes?.[0]) as ES_FIELD_TYPES;
+
+    if (fieldType === undefined) {
+      continue;
+    }
+
+    if (Boolean(dataViewField.timeSeriesMetric)) {
+      metricFields.push({
+        index: column.meta?.index ?? index,
+        name: columnName,
+        type: fieldType,
+        instrument: dataViewField.timeSeriesMetric,
+        dimensions: [],
+      });
+    } else if (Boolean(dataViewField.timeSeriesDimension) || DIMENSION_TYPES.includes(fieldType)) {
+      dimensions.push({
+        name: columnName,
+        type: fieldType,
+      });
+    }
+  }
+
+  return {
+    metricFields,
+    dimensions,
+  };
+};
+
+export interface RowMappings {
+  sampleRowByMetric: Map<string, DatatableRow>;
+  fieldSpecsByRow: WeakMap<DatatableRow, MetricField>;
+}
+
+/**
+ * Creates a map of metric field name → (value → row).
+ * @param rows - The rows to use for the sample rows.
+ * @param fieldSpecs - The field specs to use for the sample rows.
+ * @returns The sample rows by metric field name.
+ */
+export const createSampleRowByMetric = ({
+  rows = [],
+  fieldSpecs,
+}: {
+  rows: DatatableRow[];
+  fieldSpecs: MetricField[];
+}): RowMappings => {
+  const sampleRowByMetric = new Map<string, DatatableRow>();
+  const specByRow = new WeakMap<DatatableRow, MetricField>();
+
+  if (!rows?.length || fieldSpecs.length === 0) {
+    return { sampleRowByMetric, fieldSpecsByRow: specByRow };
+  }
+
+  const specByFieldName = new Map<string, MetricField>();
+  for (const spec of fieldSpecs) {
+    specByFieldName.set(spec.name, spec);
+  }
+
+  const metricFieldNames = [...specByFieldName.keys()];
+  const pendingSamples = new Set(fieldSpecs.map((s) => s.name));
+
+  for (const row of rows) {
+    for (const fieldName of metricFieldNames) {
```

Comment:
Right now we walk `metricFieldNames` for each row. We could dump these field names to a `Set` outside the loop, and instead walk only the field names on the row. For example:

```typescript
const metricFieldNamesSet = new Set([...specByFieldName.keys()]);

for (const row of rows) {
  for (const fieldName in row) {
    if (metricFieldNamesSet.has(fieldName) && hasValue(row[fieldName])) {
      // internals remain the same
    }
  }
}
```

It's still exponential time complexity but we would touch only a subset of `metricFieldNames` for the inner loop.